src/sidebar/services/frame-sync.js

```diff
@@ -235,17 +236,30 @@ export class FrameSyncService {
       });
     };
 
-    const discovery = new Discovery(window, { server: true });
-    discovery.startDiscovery(
-      (source, origin, token) =>
-        this._bridge.createChannel({ source, origin, token }),
-      [
-        // Ping the host frame which is in most cases also the only guest.
-        window.parent,
-      ]
-    );
     this._bridge.onConnect(addFrame);
 
+    // Listen for messages from new guest frames that want to connect.
+    //
+    // The message will include a `MessagePort` to use for communication with
+    // the guest. Communication with the host currently relies on the host
+    // frame also always being a guest frame.
+    this._window.addEventListener('message', e => {
+      if (e.data?.type !== 'hypothesisGuestReady') {
+        return;
+      }
+      const port = /** @type {unknown} */ (e.data.port);
```

Comment:
There is no need to transferred the port on the `data`. You can reach directly from `e.ports[0]`.

https://github.com/mdn/dom-examples/blob/92687307cb13a922e4a3ef1cc16d36fe64fcec28/channel-messaging-basic/page2.html#L24

Reply:
Ah, thanks for the tip.